# This pre-lecture lab is to help you:
* **Get familir with Pandas library**
* **Experiment with Pandas APIs**
* **Read about Pandas**
  * Pandas-Python Data Analytics Library at https://pandas.pydata.org/
  * A Guide to Pandas and Matplotlis for Data Exploration at https://towardsdatascience.com/a-guide-to-pandas-and-matplotlib-for-data-exploration-56fad95f951c
  * Plotting a time series heat map with Pandas at https://jonisalonen.com/2019/plotting-a-time-series-heat-map-with-pandas/

# **What is Pandas?**

* Pandas is a Python library used for working with data sets.

* It has functions for analyzing, cleaning, exploring, and manipulating data.

* The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

* Pandas is open source. The source code for Pandas is located at this github repository https://github.com/pandas-dev/pandas

# **Why Use Pandas?**

* Pandas allows us to analyze big data and make conclusions based on statistical theories.

* Pandas can clean messy data sets, and make them readable and relevant.

* Relevant data is very important in data science.

# **What Can Pandas Do?**

Pandas gives you answers about the data. Like:

* *Is there a correlation between two or more columns?*
* *What is average value?*
* *Max value?*
* *Min value?*

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

# **Import Pandas**

Once Pandas is installed, import it in your applications by adding the <FONT COLOR="RED">**import**</FONT> keyword:

In [ ]:
import pandas

Now Pandas is imported and ready to use.

# **Example** - reading Python <FONT COLOR="RED">*dictionary*</FONT> to Pandas *DataFrame*

In [ ]:
import pandas as pd

mydataset = {
  'participant': ["p001", "p001", "p003", "p004", "p005"],
  'age': [30, 27, 22, 25, 33],
  'weight': [125, 130, 133, 125, 133]
}

myDataFrame = pd.DataFrame(mydataset)

print(myDataFrame)


  participant  age  weight
0        p001   30     125
1        p001   27     130
2        p003   22     133
3        p004   25     125
4        p005   33     133


# **Pandas Series**

A Pandas <FONT COLOR="RED">Series</FONT> is like a column in a table.

It is a *one-dimensional array holding data of any type*.

In [ ]:
import pandas as pd

a = [1, 7, 2]

mySeries = pd.Series(a)

print(mySeries)

0    1
1    7
2    2
dtype: int64


In [ ]:
import pandas as pd

a = [1, 7, 2]

mySeries = pd.Series(a, index = ["idx1", "idx2", "idx3"])

print(mySeries)

idx1    1
idx2    7
idx3    2
dtype: int64


In [ ]:
print(mySeries["idx2"])

7


In [ ]:
print(mySeries[1])

7


# **Pandas DataFrames**

**What is a <FONT color="RED">DataFrame</FONT>?**

A ***Pandas DataFrame*** is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [ ]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


**Locate Row**

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the **<FONT color="RED">loc</FONT>** attribute to return one or more specified row(s)

In [ ]:
#refer to the row index:
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [ ]:
#use a list of indexes:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


Named Indexes
With the index argument, you can name your own indexes.

**Named Indexes**

With the index argument, you can name your own indexes.

In [ ]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


**Locate Named Indexes**

Use the named index in the **<FONT color="RED">loc</FONT>** attribute to return the specified row(s).

Example

In [ ]:
#refer to the named index:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


# **Pandas Read CSV**

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be one of those datasets that you will be analyzing and visualizing in HW2

In [ ]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/huizhangky/CSE590/main/hw/hw2/subject_registry.csv')

print(df.head(10))

   Site  SubjectID   ISA Actual_Visit Data_collection_started
0   100      11609   NaN           SF                3-Aug-21
1   100      12140  NP03     Complete               10-Aug-21
2   100      12120  BP03     Complete               11-Aug-21
3   100      11806  BP03     Complete               12-Aug-21
4   100      11653  BP03           ED               12-Aug-21
5   100      11530  BP03     Complete               25-Aug-21
6   102      12629  BP03           ED               14-Oct-21
7   102      12902  BP03           V3                3-Dec-21
8   103      12452  NP03           ED               23-Aug-21
9   103      12638  BP03           ED                7-Oct-21


Print the last 5 rows of the DataFrame:

In [ ]:
print(df.tail())

     Site  SubjectID   ISA Actual_Visit Data_collection_started
284   147      12840  OA02           V3               30-Nov-21
285   147      12821  BP03           V3               17-Nov-21
286   147      12972   NaN           V2               23-Dec-21
287   147      12541  NP03     Complete               14-Sep-21
288   147      12897  OA02           V3                2-Dec-21


# Info About the Data

The DataFrames object has a method called info(), that gives you more information about the data set.

In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 289 entries, 0 to 288
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Site                     289 non-null    int64 
 1   SubjectID                289 non-null    int64 
 2   ISA                      226 non-null    object
 3   Actual_Visit             289 non-null    object
 4   Data_collection_started  289 non-null    object
dtypes: int64(2), object(3)
memory usage: 11.4+ KB
None


# Result Explained

**The result tells us there are <FONT color="RED">289 rows and 6 columns</FONT>:**

In [ ]:
df.dtypes

Site                        int64
SubjectID                   int64
ISA                        object
Actual_Visit               object
Data_collection_started    object
dtype: object

# Convert Columns to the Desired Types

Now, we would like to make the **"Data_collection_started"** column a DateTime type

In [ ]:
df['Data_collection_started'] = pd.to_datetime(df['Data_collection_started'])

In [ ]:
df.dtypes

Site                                int64
SubjectID                           int64
ISA                                object
Actual_Visit                       object
Data_collection_started    datetime64[ns]
dtype: object

In [ ]:
df['ISA'] = df['ISA'].astype(str)
df['Actual_Visit '] = df['Actual_Visit'].astype(str)

df.dtypes

Site                                int64
SubjectID                           int64
ISA                                object
Actual_Visit                       object
Data_collection_started    datetime64[ns]
Actual_Visit                       object
dtype: object

# **Now let us query from the DataFrame**

***Let us filter out participants who belongs to BPO3 cohort***

In [ ]:
bp03_df = df.query('ISA == "BP03"')

In [ ]:
print(bp03_df)

     Site  SubjectID   ISA Actual_Visit Data_collection_started Actual_Visit 
2     100      12120  BP03     Complete              2021-08-11      Complete
3     100      11806  BP03     Complete              2021-08-12      Complete
4     100      11653  BP03           ED              2021-08-12            ED
5     100      11530  BP03     Complete              2021-08-25      Complete
6     102      12629  BP03           ED              2021-10-14            ED
..    ...        ...   ...          ...                     ...           ...
277   147      12584  BP03     Complete              2021-09-27      Complete
278   147      12595  BP03     Complete              2021-09-23      Complete
279   147      12572  BP03           ED              2021-10-07            ED
282   147      12654  BP03     Complete              2021-10-21      Complete
285   147      12821  BP03           V3              2021-11-17            V3

[132 rows x 6 columns]


***Let us further filter out participants who have NOT completed the study***

In [ ]:
bp03_completed_df = bp03_df.query('Actual_Visit == "Complete"')

In [ ]:
print(bp03_completed_df)

     Site  SubjectID   ISA Actual_Visit Data_collection_started Actual_Visit 
2     100      12120  BP03     Complete              2021-08-11      Complete
3     100      11806  BP03     Complete              2021-08-12      Complete
5     100      11530  BP03     Complete              2021-08-25      Complete
12    106      12401  BP03     Complete              2021-08-09      Complete
13    106      12328  BP03     Complete              2021-08-09      Complete
..    ...        ...   ...          ...                     ...           ...
265   142      12406  BP03     Complete              2021-10-14      Complete
268   142      12417  BP03     Complete              2021-10-19      Complete
277   147      12584  BP03     Complete              2021-09-27      Complete
278   147      12595  BP03     Complete              2021-09-23      Complete
282   147      12654  BP03     Complete              2021-10-21      Complete

[70 rows x 6 columns]


**Now let us print out the participant ids**

In [ ]:
participant_list = df['SubjectID'].tolist()

print(participant_list)

[11609, 12140, 12120, 11806, 11653, 11530, 12629, 12902, 12452, 12638, 12795, 12261, 12401, 12328, 12518, 12489, 12507, 12606, 12473, 12678, 12471, 12673, 12741, 12711, 12754, 12721, 12392, 12756, 12790, 12855, 12761, 12763, 11624, 12212, 12213, 12192, 12235, 12259, 12312, 12456, 12501, 12517, 12544, 12552, 12565, 12587, 12585, 12668, 12709, 12720, 12706, 12848, 12832, 12462, 12893, 12807, 12985, 12677, 12918, 12225, 12475, 12592, 12641, 12644, 12671, 12696, 12688, 12690, 12731, 12672, 12785, 12849, 12775, 12824, 12174, 12063, 12100, 12050, 12478, 12136, 12375, 12395, 12578, 12472, 12270, 12730, 12729, 12755, 12281, 12591, 12830, 12783, 12928, 12773, 12771, 12805, 12792, 12804, 12175, 12539, 12469, 12480, 12500, 12463, 12561, 12490, 12058, 12877, 12615, 12878, 12740, 12506, 12753, 12719, 12784, 12939, 12856, 11611, 11915, 12458, 11911, 11954, 12504, 12498, 12820, 12597, 12599, 12623, 12627, 12779, 12838, 11976, 11701, 11902, 11985, 11568, 11980, 12533, 12189, 12369, 12549, 12604, 12302

In [ ]:
import pandas as pd
pd.to_datetime("2021-05-18")

Timestamp('2021-05-18 00:00:00')